src/lib/core/prbuf.c

```diff
@@ -0,0 +1,359 @@
+ */
+
+/** A mark of unused space in the buffer: trash is located after this point. */
+static const uint32_t prbuf_end_position = (uint32_t)(-1);
+
+/** Before storing a data in the buffer we place its size (i.e. header). */
+static const size_t record_size_overhead = sizeof(struct prbuf_record);
+
+/** Real size of allocation is (data size + record's header). */
+static uint32_t
+prbuf_record_alloc_size(size_t size)
+{
+	return size + offsetof(struct prbuf_record, data);
+}
+
+/** Returns pointer to the next byte after end of given buffer. */
+static char *
+prbuf_linear_end(struct prbuf *buf)
+{
+	return (char *)buf->header + buf->header->size;
+}
+
+/** Returns pointer to the first writable byte of given buffer. */
+static char *
+prbuf_linear_begin(struct prbuf *buf)
+{
+	return (char *)buf->header + sizeof(struct prbuf_header);
+}
+
+/** Returns pointer to the next byte after the last written record. */
+static char *
+prbuf_current_raw(struct prbuf *buf)
+{
+	return prbuf_linear_begin(buf) + buf->header->end;
+}
+
+static struct prbuf_record *
+prbuf_current_record(struct prbuf *buf)
+{
+	return (struct prbuf_record *)prbuf_current_raw(buf);
+}
+
+/** Returns first (in historical sense) record. */
+static struct prbuf_record *
+prbuf_first_record(struct prbuf *buf)
+{
+	assert(buf->header->begin != prbuf_end_position);
+	char *first_ptr = prbuf_linear_begin(buf) + buf->header->begin;
+	return (struct prbuf_record *)first_ptr;
+}
+
+/** Calculate offset from the buffer's start to the given entry. */
+static uint32_t
+prbuf_record_offset(struct prbuf *buf, struct prbuf_record *record)
+{
+	assert((char *)record >= prbuf_linear_begin(buf));
+	return (uint32_t)((char *)record - prbuf_linear_begin(buf));
+}
+
+/** Returns true in case buffer has at least @a size bytes until
+ * its linear end.
+ */
+static bool
+prbuf_has_before_end(struct prbuf *buf, uint32_t size)
+{
+	assert(prbuf_linear_end(buf) >= prbuf_current_raw(buf));
+	if ((uint32_t)(prbuf_linear_end(buf) - prbuf_current_raw(buf)) >= size)
+		return true;
+	return false;
+}
+
+static uint32_t
+prbuf_size(struct prbuf *buf)
+{
+	return buf->header->size - sizeof(struct prbuf_header);
+}
+
+void
+prbuf_create(struct prbuf *buf, void *mem, size_t size)
+{
+	assert(size > sizeof(struct prbuf_header));
+#ifndef NDEBUG
+	memset(mem, '#', size);
+#endif
+	buf->header = (struct prbuf_header *)mem;
+	buf->header->end = 0;
+	buf->header->begin = 0;
+	buf->header->version = prbuf_version;
+	buf->header->size = size;
+}
+
+/** Returns true in case buffer contains no entries. */
+static bool
+prbuf_is_empty(struct prbuf *buf)
+{
+	return buf->header->end == 0;
+}
+
+/**
+ * Assuming @a record pointing to the valid buffer position, move it to the
+ * next entry. In case of current entry is broken (contains wrong size or
+ * points out of the buffer) function returns -1.
+ */
+static int
+prbuf_record_next(struct prbuf *buf, struct prbuf_record **record)
```

Comment:
Nit: Please rename to `prbuf_next_record`, for consistency with `prbuf_skip_record`, `prbuf_prepare_record`, etc.

Reply:
Ok, renamed.